java.lang.ClassCastException: kotlin.coroutines.jvm.internal.CompletedContinuation cannot be cast to kotlinx.coroutines.internal.DispatchedContinuation

```
21:05:05	ERROR:	AndroidRuntime : kotlinx.coroutines.CoroutinesInternalError: Fatal exception in coroutines machinery for CancellableContinuation(DispatchedContinuation[Dispatchers.IO, Continuation at kotlinx.coroutines.channels.AbstractChannel.receiveCatching-JP2dKIU(AbstractChannel.kt:632)@8d8755]){Completed}@c3cb56a. Please read KDoc to 'handleFatalException' method and report this incident to maintainers
21:05:05	ERROR:	AndroidRuntime : at kotlinx.coroutines.DispatchedTask.handleFatalException(DispatchedTask.kt:144)
21:05:05	ERROR:	AndroidRuntime : at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:115)
21:05:05	ERROR:	AndroidRuntime : at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
21:05:05	ERROR:	AndroidRuntime : at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
21:05:05	ERROR:	AndroidRuntime : at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
21:05:05	ERROR:	AndroidRuntime : at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
21:05:05	ERROR:	AndroidRuntime : Caused by: java.lang.ClassCastException: kotlin.coroutines.jvm.internal.CompletedContinuation cannot be cast to kotlinx.coroutines.internal.DispatchedContinuation
21:05:05	ERROR:	AndroidRuntime : at kotlinx.coroutines.CoroutineDispatcher.releaseInterceptedContinuation(CoroutineDispatcher.kt:108)
21:05:05	ERROR:	AndroidRuntime : at kotlin.coroutines.jvm.internal.ContinuationImpl.releaseIntercepted(ContinuationImpl.kt:118)
21:05:05	ERROR:	AndroidRuntime : at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:39)
21:05:05	ERROR:	AndroidRuntime : at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
21:05:05	ERROR:	AndroidRuntime : ... 4 more
```

kotlin version: 1.5.10
coroutines version: 1.5.0

I'm not sure how to reproduce, it sometimes happens during espresso tests

@qwwdfsad it is, run it multiple times and sometimes it fails with exact classcastexception: 
to increase your chances (i don't know why) run more coroutines:
```
suspend fun runTest() = withContext(Dispatchers.Default) {
        repeat(300) { a ->
            repeat(300) { b ->
                launch {
                    runTestOnce(a + 50, b + 50)
                }
            }
        }
    }
```

produces 
```
Exception in thread "DefaultDispatcher-worker-8" kotlinx.coroutines.CoroutinesInternalError: Fatal exception in coroutines machinery for CancellableContinuation(DispatchedContinuation[Dispatchers.Default, Continuation at kotlinx.coroutines.flow.StateFlowImpl.collect(StateFlow.kt:353)@6476f03]){Cancelled}@10ee2b07. Please read KDoc to 'handleFatalException' method and report this incident to maintainers
	at kotlinx.coroutines.DispatchedTask.handleFatalException(DispatchedTask.kt:144)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:115)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
Caused by: java.lang.ClassCastException: class kotlin.coroutines.jvm.internal.CompletedContinuation cannot be cast to class kotlinx.coroutines.internal.DispatchedContinuation (kotlin.coroutines.jvm.internal.CompletedContinuation and kotlinx.coroutines.internal.DispatchedContinuation are in unnamed module of loader 'app')
	at kotlinx.coroutines.CoroutineDispatcher.releaseInterceptedContinuation(CoroutineDispatcher.kt:108)
	at kotlin.coroutines.jvm.internal.ContinuationImpl.releaseIntercepted(ContinuationImpl.kt:118)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:39)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:104)
	... 4 more
```
in most of the runs (and it fails faster)

--- a/kotlinx-coroutines-core/common/src/CancellableContinuationImpl.kt
+++ b/kotlinx-coroutines-core/common/src/CancellableContinuationImpl.kt
@@ -91,6 +91,18 @@ internal open class CancellableContinuationImpl<in T>(
 
     public override fun initCancellability() {
         /*
+         * We could've been in the following situation:
+         * suspendCancellableCoroutineReusable {} // <- CLAIMED, trySuspend(), now blocked by OS schedulder
+         * suspendCancellableCoroutine {} // <- we are here
+         *
+         * In such rare cases, we explicitly await for reusable continuation to became available and
+         * for `isReusable` to start returning `false`.
+         */
+        if (isReusable()) {
+            (delegate as DispatchedContinuation<*>).awaitReusability()
+        }
+
+        /*
         * Invariant: at the moment of invocation, `this` has not yet
         * leaked to user code and no one is able to invoke `resume` or `cancel`
         * on it yet. Also, this function is not invoked for reusable continuations.

--- a/kotlinx-coroutines-core/common/src/CancellableContinuationImpl.kt
+++ b/kotlinx-coroutines-core/common/src/CancellableContinuationImpl.kt
@@ -91,18 +91,6 @@ internal open class CancellableContinuationImpl<in T>(
 
     public override fun initCancellability() {
         /*
-         * We could've been in the following situation:
-         * suspendCancellableCoroutineReusable {} // <- CLAIMED, trySuspend(), now blocked by OS schedulder
-         * suspendCancellableCoroutine {} // <- we are here
-         *
-         * In such rare cases, we explicitly await for reusable continuation to became available and
-         * for `isReusable` to start returning `false`.
-         */
-        if (isReusable()) {
-            (delegate as DispatchedContinuation<*>).awaitReusability()
-        }
-
-        /*
         * Invariant: at the moment of invocation, `this` has not yet
         * leaked to user code and no one is able to invoke `resume` or `cancel`
         * on it yet. Also, this function is not invoked for reusable continuations.
@@ -119,7 +107,7 @@ internal open class CancellableContinuationImpl<in T>(
         }
     }
 
-    private fun isReusable(): Boolean = delegate is DispatchedContinuation<*> && delegate.isReusable(this)
+    private fun isReusable(): Boolean = resumeMode.isReusableMode && (delegate as DispatchedContinuation<*>).isReusable()
 
     /**
      * Resets cancellability state in order to [suspendCancellableCoroutineReusable] to work.
@@ -127,7 +115,7 @@ internal open class CancellableContinuationImpl<in T>(
      */
     @JvmName("resetStateReusable") // Prettier stack traces
     internal fun resetStateReusable(): Boolean {
-        assert { resumeMode == MODE_CANCELLABLE_REUSABLE } // invalid mode for CancellableContinuationImpl
+        assert { resumeMode == MODE_CANCELLABLE_REUSABLE }
         assert { parentHandle !== NonDisposableHandle }
         val state = _state.value
         assert { state !is NotCompleted }
@@ -176,8 +164,7 @@ internal open class CancellableContinuationImpl<in T>(
      * Attempt to postpone cancellation for reusable cancellable continuation
      */
     private fun cancelLater(cause: Throwable): Boolean {
-        if (!resumeMode.isReusableMode) return false
-        // Ensure that we are postponing cancellation to the right instance
+        // Ensure that we are postponing cancellation to the right reusable instance
         if (!isReusable()) return false
         val dispatched = delegate as DispatchedContinuation<*>
         return dispatched.postponeCancellation(cause)

--- a/kotlinx-coroutines-core/common/src/internal/DispatchedContinuation.kt
+++ b/kotlinx-coroutines-core/common/src/internal/DispatchedContinuation.kt
@@ -58,16 +58,16 @@ internal class DispatchedContinuation<in T>(
      */
     private val _reusableCancellableContinuation = atomic<Any?>(null)
 
-    public val reusableCancellableContinuation: CancellableContinuationImpl<*>?
+    val reusableCancellableContinuation: CancellableContinuationImpl<*>?
         get() = _reusableCancellableContinuation.value as? CancellableContinuationImpl<*>
 
-    public fun isReusable(requester: CancellableContinuationImpl<*>): Boolean {
+    fun isReusable(requester: CancellableContinuationImpl<*>): Boolean {
         /*
          * Reusability control:
          * `null` -> no reusability at all, `false`
          * If current state is not CCI, then we are within `suspendCancellableCoroutineReusable`, true
          * Else, if result is CCI === requester, then it's our reusable continuation
-         * Identity check my fail for the following pattern:
+         * Identity check may fail for the following pattern:
          * ```
          * loop:
          * suspendCancellableCoroutineReusable { } // Reusable, outer coroutine stores the child handle
@@ -80,13 +80,12 @@ internal class DispatchedContinuation<in T>(
         return true
     }
 
-
     /**
      * Awaits until previous call to `suspendCancellableCoroutineReusable` will
      * stop mutating cached instance
      */
     public fun awaitReusability() {
-        _reusableCancellableContinuation.loop { it ->
+        _reusableCancellableContinuation.loop {
             if (it !== REUSABLE_CLAIMED) return
         }
     }

--- a/kotlinx-coroutines-core/common/src/internal/DispatchedContinuation.kt
+++ b/kotlinx-coroutines-core/common/src/internal/DispatchedContinuation.kt
@@ -58,39 +58,30 @@ internal class DispatchedContinuation<in T>(
      */
     private val _reusableCancellableContinuation = atomic<Any?>(null)
 
-    val reusableCancellableContinuation: CancellableContinuationImpl<*>?
+    private val reusableCancellableContinuation: CancellableContinuationImpl<*>?
         get() = _reusableCancellableContinuation.value as? CancellableContinuationImpl<*>
 
-    fun isReusable(requester: CancellableContinuationImpl<*>): Boolean {
+    fun isReusable(): Boolean {
         /*
+        Invariant: caller.resumeMode.isReusableMode
          * Reusability control:
          * `null` -> no reusability at all, `false`
-         * If current state is not CCI, then we are within `suspendCancellableCoroutineReusable`, true
-         * Else, if result is CCI === requester, then it's our reusable continuation
-         * Identity check may fail for the following pattern:
-         * ```
-         * loop:
-         * suspendCancellableCoroutineReusable { } // Reusable, outer coroutine stores the child handle
-         * suspendCancellableCoroutine { } // **Not reusable**, handle should be disposed after {}, otherwise
-         * it will leak because it won't be freed by `releaseInterceptedContinuation`
-         * ```
+         * anything else -> reusable.
          */
-        val value = _reusableCancellableContinuation.value ?: return false
-        if (value is CancellableContinuationImpl<*>) return value === requester
-        return true
+        return _reusableCancellableContinuation.value != null
     }
 
     /**
      * Awaits until previous call to `suspendCancellableCoroutineReusable` will
      * stop mutating cached instance
      */
-    public fun awaitReusability() {
+    fun awaitReusability() {
         _reusableCancellableContinuation.loop {
             if (it !== REUSABLE_CLAIMED) return
         }
     }
 
-    public fun release() {
+    fun release() {
         /*
          * Called from `releaseInterceptedContinuation`, can be concurrent with
          * the code in `getResult` right after `trySuspend` returned `true`, so we have

--- a/kotlinx-coroutines-core/jvm/test/ReusableContinuationStressTest.kt
+++ b/kotlinx-coroutines-core/jvm/test/ReusableContinuationStressTest.kt
@@ -1,0 +1,41 @@
+/*
+ * Copyright 2016-2021 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.coroutines
+
+import kotlinx.coroutines.flow.*
+import org.junit.*
+
+class ReusableContinuationStressTest : TestBase() {
+
+    private val iterations = 1000 * stressTestMultiplierSqrt
+
+    @Test // Originally reported by @denis-bezrukov in #2736
+    fun testDebounceWithStateFlow() = runBlocking<Unit> {
+        withContext(Dispatchers.Default) {
+            repeat(iterations) {
+                launch { // <- load the dispatcher and OS scheduler
+                    runStressTestOnce(1, 1)
+                }
+            }
+        }
+    }
+
+    private suspend fun runStressTestOnce(delay: Int, debounce: Int) = coroutineScope {
+        val stateFlow = MutableStateFlow(0)
+        val emitter = launch {
+            repeat(1000) { i ->
+                stateFlow.emit(i)
+                delay(delay.toLong())
+            }
+        }
+        var last = 0
+        stateFlow.debounce(debounce.toLong()).take(100).collect { i ->
+            if (i - last > 100) {
+                last = i
+            }
+        }
+        emitter.cancel()
+    }
+}

